parsec/data_distribution.c

```diff
@@ -53,22 +53,46 @@ parsec_data_dist_fini(void)
 int
 parsec_dc_register_id(parsec_dc_t* dc, parsec_dc_key_t key)
 {
-    assert(NULL != parsec_dc_hash_table);
+    if( NULL == parsec_dc_hash_table ) {
+        /* Do not call this function before initializing the runtime */
+        return PARSEC_ERROR;
+    }
+#if defined(PARSEC_DEBUG_PARANOID)
+    parsec_dc_t* registered_dc = (parsec_dc_t*)parsec_hash_table_find(parsec_dc_hash_table, key);
+    if( dc == registered_dc ) {
+        /* complain but allow the code to continue */
+        parsec_inform("Re-registering a data collection (%p) with the same key (%lld). This operation "
+                      "is unnecessary and can lead to unexpected consequences. Please correct "
+                      "the code", dc, (long long)key);
+    } else if( NULL != registered_dc ) {
+        PARSEC_DEBUG_VERBOSE(10, parsec_debug_output,
+                             "Registering a data collection with an already existing key ID (%lld). "
+                             "The ID keys must be unique to avoid collisions between different "
+                             "data collections.", (long long)key);
+        return PARSEC_ERROR;
+    }
+#endif  /* defined(PARSEC_DEBUG_PARANOID) */
     dc->ht_item.key = key;
     dc->dc_id = key;
     parsec_hash_table_insert(parsec_dc_hash_table, &dc->ht_item);
     /* We do not have any function to check if there is a collision
      * so for now we replace the old entry if there is request for
      * a new registration with same id.
      */
+    parsec_inform("register dc_t %p with key %lld", dc, (long long)key);
```

Comment:
This is way too high a debug level for this, should use parsec_debug_verbose(...)

Reply:
This is a real issue, it might or might not work we have no way to know if there will be collision or not between different data collections. So I would rather have this well visible, such that an educated user can fix their code (once it gets annoyed by the output).

Reply:
looking at the code, I think you are referring to the line 64: registering the same key twice (this is abnormal but harmless, `parsec_info` is adequate), on line 68, the `DEBUG_VERBOSE` should be promoted to a `parsec_warning` because this is a real coding error that will cause problems, this line is unconditionally printing the registration, this is too verbose.

Reply:
ok, I will make the change. It is unfortunate that the warnings are only delivered when in heavy debug mode.

Reply:
See 52caa8bee925eac4c3876079444690662393f69a